[WPE] WPEPlatform: implement EventSenderProxy touch methods

carlosgcampos opened the pull request:
#### 14f5637a0139d0acb53a8a1bd2845eb10ecb8ea1
<pre>
[WPE] WPEPlatform: implement EventSenderProxy touch methods
<a href="https://bugs.webkit.org/show_bug.cgi?id=292766">https://bugs.webkit.org/show_bug.cgi?id=292766</a>

Reviewed by Alejandro G. Castro.

Also fix a bug I found while running tests in touchPointsForEvent() that
we were passing the wrong WPEEvent.

* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
(WKWPE::ViewPlatform::touchPointsForEvent):
* Tools/WebKitTestRunner/wpe/EventSenderProxyClientWPE.cpp:
(WTR::EventSenderProxyClientWPE::addTouchPoint):
(WTR::EventSenderProxyClientWPE::updateTouchPoint):
(WTR::EventSenderProxyClientWPE::releaseTouchPoint):
(WTR::EventSenderProxyClientWPE::cancelTouchPoint):
(WTR::EventSenderProxyClientWPE::clearTouchPoints):
(WTR::EventSenderProxyClientWPE::touchStart):
(WTR::EventSenderProxyClientWPE::touchMove):
(WTR::EventSenderProxyClientWPE::touchEnd):
(WTR::EventSenderProxyClientWPE::touchCancel):
(WTR::EventSenderProxyClientWPE::setTouchModifier):
* Tools/WebKitTestRunner/wpe/EventSenderProxyClientWPE.h:

Canonical link: <a href="https://commits.webkit.org/294892@main">https://commits.webkit.org/294892@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d858807994bb72db1ba7908f350eeb793c3a20e4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103290 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22966 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13287 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108463 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53933 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105329 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23305 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31517 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78495 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35422 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106296 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18063 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93182 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58828 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17899 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11224 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53289 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87704 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11286 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110837 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30428 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22468 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87492 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30793 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89380 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87128 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31990 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9708 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24751 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16772 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30357 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35676 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30163 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33490 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31725 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

csaavedra commented on Tools/WebKitTestRunner/wpe/EventSenderProxyClientWPE.cpp:
```diff
@@ -303,6 +303,114 @@ void EventSenderProxyClientWPE::keyDown(WKStringRef keyRef, double time, WKEvent
+            continue;
+
+        point.state = TouchPoint::State::Stationary;
+        auto* event = wpe_event_touch_new(WPE_EVENT_TOUCH_DOWN, view, WPE_INPUT_SOURCE_TOUCHSCREEN, secToMsTimestamp(time), static_cast<WPEModifiers>(m_touchModifiers), point.id, point.x, point.y);
+        wpe_view_event(view, event);
+        wpe_event_unref(event);
```
I imagine we don't have some unique pointer support for these classes (yet) to avoid the manual ref/unreffing?

carlosgcampos replied:
We have it in WebKit layer, but not in WTF because wpe platform api is UI process only lib

spenap commented on Tools/WebKitTestRunner/wpe/EventSenderProxyClientWPE.cpp:
```diff
@@ -303,6 +303,114 @@ void EventSenderProxyClientWPE::keyDown(WKStringRef keyRef, double time, WKEvent
+
+void EventSenderProxyClientWPE::updateTouchPoint(int index, int x, int y, double)
+{
+    ASSERT(index >= 0 && static_cast<size_t>(index) <= m_touchPoints.size());
```
```suggestion
    ASSERT(index >= 0 && static_cast<size_t>(index) < m_touchPoints.size());
```

if `index == m_touchPoints.size()` we'd be accessing out of bounds, wouldn't we?

carlosgcampos replied:
Right, I copied the asserts from the libwpe implementation, that should be fixed in a different patch

spenap commented on Tools/WebKitTestRunner/wpe/EventSenderProxyClientWPE.cpp:
```diff
@@ -303,6 +303,114 @@ void EventSenderProxyClientWPE::keyDown(WKStringRef keyRef, double time, WKEvent
+
+void EventSenderProxyClientWPE::releaseTouchPoint(int index, double)
+{
+    ASSERT(index >= 0 && static_cast<size_t>(index) <= m_touchPoints.size());
```
```suggestion
    ASSERT(index >= 0 && static_cast<size_t>(index) < m_touchPoints.size());
```

ditto

spenap commented on Tools/WebKitTestRunner/wpe/EventSenderProxyClientWPE.cpp:
```diff
@@ -303,6 +303,114 @@ void EventSenderProxyClientWPE::keyDown(WKStringRef keyRef, double time, WKEvent
+
+void EventSenderProxyClientWPE::cancelTouchPoint(int index, double)
+{
+    ASSERT(index >= 0 && static_cast<size_t>(index) <= m_touchPoints.size());
```
```suggestion
    ASSERT(index >= 0 && static_cast<size_t>(index) < m_touchPoints.size());
```

ditto

spenap reviewed (commented):
Some feedback on the assertions. And maybe as a follow-up (I'm happy to volunteer) we could refactor the `touch*` methods so that they use a common method underneath

alexgcastro commented on Tools/WebKitTestRunner/wpe/EventSenderProxyClientWPE.cpp:
```diff
@@ -303,6 +303,114 @@ void EventSenderProxyClientWPE::keyDown(WKStringRef keyRef, double time, WKEvent
+void EventSenderProxyClientWPE::addTouchPoint(int x, int y, double)
+{
+    uint32_t id = 0;
+    for (const auto& point : m_touchPoints) {
+        if (point.id == id)
+            id++;
+    }
```
Can this fail if we remove previously added points from the array and we try to add a new points?

carlosgcampos replied:
No, it's ok to reuse ids.

alexgcastro reviewed (approved):
LGTM!

carlosgcampos committed:
[WPE] WPEPlatform: implement EventSenderProxy touch methods
https://bugs.webkit.org/show_bug.cgi?id=292766

Reviewed by Alejandro G. Castro.

Also fix a bug I found while running tests in touchPointsForEvent() that
we were passing the wrong WPEEvent.

* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
(WKWPE::ViewPlatform::touchPointsForEvent):
* Tools/WebKitTestRunner/wpe/EventSenderProxyClientWPE.cpp:
(WTR::EventSenderProxyClientWPE::addTouchPoint):
(WTR::EventSenderProxyClientWPE::updateTouchPoint):
(WTR::EventSenderProxyClientWPE::releaseTouchPoint):
(WTR::EventSenderProxyClientWPE::cancelTouchPoint):
(WTR::EventSenderProxyClientWPE::clearTouchPoints):
(WTR::EventSenderProxyClientWPE::touchStart):
(WTR::EventSenderProxyClientWPE::touchMove):
(WTR::EventSenderProxyClientWPE::touchEnd):
(WTR::EventSenderProxyClientWPE::touchCancel):
(WTR::EventSenderProxyClientWPE::setTouchModifier):
* Tools/WebKitTestRunner/wpe/EventSenderProxyClientWPE.h:

Canonical link: https://commits.webkit.org/294892@main
